demos: add cert-manager demo (#311)
openservicemesh/osm-docs: content/docs/demos/cert-manager_integration.md

---
title: "Cert-manager Certificate Provider"
description: "Using cert-manager as a certificate provider"
type: docs
weight: 30
---

This guide demonstrates the usage of [cert-manager][1] as a certificate provider to manage and issue certificates in OSM.

## Prerequisites

- Kubernetes cluster running Kubernetes v1.19.0 or greater.

> - Have OSM installed.

Reviewer: You have a note below stating that cert-manager must be installed and running before osm is installed to use cert-manager as its certificate provider. Maybe remove this from the pre-requisites ?

- Have `kubectl` available to interact with the API server.

> - Have `osm` CLI available for managing the service mesh.

Reviewer: ```suggestion
- Have `osm` CLI available for installing and managing the service mesh.
```

    ```

    Confirm the pods are ready and running in the `cert-manager` namespace.
    ```console
    kubectl get pod -n cert-manager
    NAME                                      READY   STATUS    RESTARTS   AGE
    cert-manager-55658cdf68-pdnzg             1/1     Running   0          2m33s
    cert-manager-cainjector-967788869-prtjq   1/1     Running   0          2m33s
    cert-manager-webhook-6668fbb57d-vzm4j     1/1     Running   0          2m33s
    ```

1. Configure `cert-manager` `Issuer` and `Certificate` resources required by `cert-manager` to be able to issue certificates in OSM. These resources must be created in the namespace where OSM will be installed later.
    > Note: `cert-manager` must first be installed, with an issuer ready, before OSM can be installed using `cert-manager` as the certificate provider.

    ```bash
    # Create the namespace where OSM will be installed
    export osm_namespace=osm-system # Replace osm-system with the namespace where OSM is installed
    kubectl create namespace "$osm_namespace"

    # Create Issuer and Certificate resources
    kubectl apply -f - <<EOF
    apiVersion: cert-manager.io/v1
    kind: Issuer
    metadata:

>       name: selfsigned

Reviewer 2: I think it would be helpful to explain that we are using the selfsigned issuer to bootstrap a custom root certificate.

      namespace: "$osm_namespace"
    spec:
      selfSigned: {}
    ---
    apiVersion: cert-manager.io/v1
    kind: Certificate
    metadata:
      name: osm-ca
      namespace: "$osm_namespace"
    spec:
      isCA: true

>       duration: 87600h # 365 days
> 
>       secretName: osm-ca-bundle

Reviewer 2: ```suggestion
      duration: 87600h # 365 days

      secretName: osm-ca-bundle
```
```suggestion
      duration: 87600h # 365 days
      secretName: osm-ca-bundle
```

      name: osm-ca
      namespace: "$osm_namespace"
    spec:
      ca:
        secretName: osm-ca-bundle
    EOF
    ```

1. Confirm the `osm-ca-bundle` CA secret is created by `cert-manager` in OSM's namespace.
    ```console
    $ kubectl get secret osm-ca-bundle -n "$osm_namespace"
    NAME            TYPE                DATA   AGE
    osm-ca-bundle   kubernetes.io/tls   3      84s
    ```

    The CA certificate saved in this secret will be used by OSM upon install to bootstrap its ceritifcate provider utility.

1. Install OSM with its certificate provider kind set to `cert-manager`.
    ```bash
    osm install --set osm.certificateProvider.kind="cert-manager"
    ```

    Confirm the OSM control plane pods are ready and running.
    ```console
    $ kubectl get pod -n "$osm_namespace"
    NAME                              READY   STATUS    RESTARTS   AGE
    osm-bootstrap-7ddc6f9b85-k8ptp    1/1     Running   0          2m52s
    osm-controller-79b777889b-mqk4g   1/1     Running   0          2m52s
    osm-injector-5f96468fb7-p77ps     1/1     Running   0          2m52s
    ```

1. Enable permissive traffic policy mode to set up automatic application connectivity.

>     > Note: this is not required to use `cert-manager` but simplifies the demo by not requiring explicit traffic policies for application connectivity.

Reviewer: ```suggestion
    > Note: this is not a requirement to use `cert-manager` but simplifies the demo by not requiring explicit traffic policies for application connectivity.
```
